# Anem a resoldre l'exemplre de viatjar per les ciutats de Romania

Has de buscar el camí més curt per anar de Arad a Bucharest. Per això has de crear un graf amb les ciutats i les distàncies entre elles. Les distàncies les pots trobar al diccionari `distancies` que ja està creat.

Has d'utilitzar la metodologia que ja hem vist a classe, creant l'estat `EstatRomania` i el problema ProblemaRomania. Després has de resoldre el problema amb ucs i A*.

Et proporcionem la classe Problema, que ja hem utilitzat, el diccionari de distàncies i un altre diccionari amb les distàncies en línia recta entre les ciutats i Bucharest.


In [1]:
distancies = {
    'Arad': {'Zerind': 75, 'Sibiu': 140, 'Timisoara': 118},
    'Zerind': {'Arad': 75, 'Oradea': 71},
    'Oradea': {'Zerind': 71, 'Sibiu': 151},
    'Sibiu': {'Arad': 140, 'Oradea': 151, 'Fagaras': 99, 'Rimnicu Vilcea': 80},
    'Timisoara': {'Arad': 118, 'Lugoj': 111},
    'Lugoj': {'Timisoara': 111, 'Mehadia': 70},
    'Mehadia': {'Lugoj': 70, 'Drobeta': 75},
    'Drobeta': {'Mehadia': 75, 'Craiova': 120},
    'Craiova': {'Drobeta': 120, 'Rimnicu Vilcea': 146, 'Pitesti': 138},
    'Rimnicu Vilcea': {'Sibiu': 80, 'Craiova': 146, 'Pitesti': 97},
    'Fagaras': {'Sibiu': 99, 'Bucharest': 211},
    'Pitesti': {'Rimnicu Vilcea': 97, 'Craiova': 138, 'Bucharest': 101},
    'Bucharest': {'Fagaras': 211, 'Pitesti': 101, 'Giurgiu': 90},
    'Giurgiu': {'Bucharest': 90}
}

distancies_recta = {
    'Arad': 366,
    'Zerind': 374,
    'Oradea': 380,
    'Sibiu': 253,
    'Timisoara': 329,
    'Lugoj': 244,
    'Mehadia': 241,
    'Drobeta': 242,
    'Craiova': 160,
    'Rimnicu Vilcea': 193,
    'Fagaras': 176,
    'Pitesti': 100,
    'Bucharest': 0,
    'Giurgiu': 77
}


In [2]:
class Problema(object):
    """Aquesta és la classe abstracta per a un problema formal. Una nova àrea crea una subclasse d'aquesta, sobrescrivint `accions` i `accio`, i potser altres mètodes.
    L'heurística per defecte és 0 i el cost d'acció per defecte és 1 per a tots els estats.
    Quan crees una instància d'una subclasse, especifica els estats `inicial` i `final`
    (o proporciona un mètode `es_resultat`) i potser altres arguments de paraula clau per a la subclasse."""

    def __init__(self, inicial=None, final=None, **kwds):
        self.__dict__.update(inicial=inicial, final=final, **kwds)

    def accions(self, state):         raise NotImplementedError

    def accio(self, state, action):   raise NotImplementedError

    def es_resultat(self, state):     return state == self.final

    def cost_accio(self, s, a, s1):   return 1

    def h(self, estat):               return 0

    def __str__(self):
        return '{}({!r}, {!r})'.format(
            type(self).__name__, self.inicial, self.final)

In [3]:
from dataclasses import dataclass


@dataclass
class EstatRomania(object):
    """Aquesta classe representa un estat del problema de Romania. L'estat és una ciutat."""

    ciutat: str
    pare: 'EstatRomania' = None

    def __repr__(self):
        return 'EstatRomania({!r})'.format(self.ciutat)

    def __str__(self):
        return '{}'.format(self.ciutat)

    def __eq__(self, other):
        return self.ciutat == other.ciutat

    def __hash__(self):
        return hash(self.ciutat)

    def __lt__(self, other):
        return self.ciutat < other.ciutat

In [4]:
class ProblemaRomania(Problema):
    """Aquesta classe representa el problema de Romania. L'estat és una ciutat."""

    def __init__(self, inicial, final):
        super().__init__(inicial, final)

    def accions(self, estat):
        """Retorna una llista d'accions possibles a l'estat actual."""
        return list(distancies[estat.ciutat].keys())

    def accio(self, estat, accio):
        """Retorna l'estat que resulta d'aplicar l'accio a l'estat."""
        return EstatRomania(accio, estat)

    def cost_accio(self, estat, accio, estat1):
        """Retorna el cost d'aplicar l'accio a l'estat."""
        return distancies[estat.ciutat][accio]

    def h(self, estat):
        """Retorna una estimació heurística del cost per anar de l'estat a l'estat final."""
        return distancies_recta[estat.ciutat]

In [5]:
import heapq

def ucs(problema):
    """Algorisme de cerca uniforme. Retorna una llista amb el camí des de l'estat inicial fins a l'estat final."""
    frontera = [(0, problema.inicial)]
    visitats = set()
    while frontera:
        cost, estat = heapq.heappop(frontera)
        if problema.es_resultat(estat):
            print('Visitats:', len(visitats))
            return estat
        if estat not in visitats:
            visitats.add(estat)
            for accio in problema.accions(estat):
                estat1 = problema.accio(estat, accio)
                cost1 = cost + problema.cost_accio(estat, accio, estat1)
                heapq.heappush(frontera, (cost1, estat1))
        

In [6]:
def extreu_ruta(estat):
    """Retorna una llista amb els estats de la ruta des de l'estat inicial fins a l'estat final."""
    ruta = []
    while estat.pare:
        ruta.append(estat)
        estat = estat.pare
    ruta.append(estat)
    ruta.reverse()
    return ruta

def mostra_ruta(estat):
    """Mostra la ruta des de l'estat inicial fins a l'estat final."""
    ruta = extreu_ruta(estat)
    cost = 0
    for estat in ruta:
        if estat.pare:
            print(f"{estat.pare} -> {estat}  ({distancies[estat.pare.ciutat][estat.ciutat]})")
            cost += distancies[estat.pare.ciutat][estat.ciutat]
        else:
            print(estat)

    print(f"Cost: {cost}")

In [7]:
problema = ProblemaRomania(EstatRomania('Arad'), EstatRomania('Bucharest'))
estat = ucs(problema)

mostra_ruta(estat)

Visitats: 12
Arad
Arad -> Sibiu  (140)
Sibiu -> Rimnicu Vilcea  (80)
Rimnicu Vilcea -> Pitesti  (97)
Pitesti -> Bucharest  (101)
Cost: 418


In [ ]:
def a_estrella(problema):
    """Algorisme de cerca A*. Retorna una llista amb el camí des de l'estat inicial fins a l'estat final."""
    frontera = [(problema.h(problema.inicial), 0, problema.inicial)]
    visitats = set()
    while frontera:
        _, cost_actual, estat = heapq.heappop(frontera)

        if estat in visitats:
            continue

        visitats.add(estat)

        if problema.es_resultat(estat):
            print('Visitats:', len(visitats))
            return estat

        for accio in problema.accions(estat):
            nou_estat = problema.accio(estat, accio)
            nou_cost = cost_actual + problema.cost_accio(estat, accio, nou_estat)
            prioritat = nou_cost + problema.h(nou_estat)

            heapq.heappush(frontera, (prioritat, nou_cost, nou_estat))


In [13]:
problema = ProblemaRomania(EstatRomania('Arad'), EstatRomania('Bucharest'))
estat = a_estrella(problema)

mostra_ruta(estat)


Visitats: 6
Arad
Arad -> Sibiu  (140)
Sibiu -> Rimnicu Vilcea  (80)
Rimnicu Vilcea -> Pitesti  (97)
Pitesti -> Bucharest  (101)
Cost: 418


# Exercici: A* ponderat

A* utilitza la prioritat $f(n) = g(n) + h(n)$. Ara investigarem què passa si donem més pes a l'heurística:

$$f_w(n) = g(n) + w h(n)$$

- $w = 0$: cerca de cost uniforme.
- $w = 1$: A* estàndard.
- $w > 1$: A* ponderat; prioritza més l'aproximació al destí.

Copia `a_estrella` a la cel·la següent amb el nom `a_estrella_ponderada(problema, pes=1)`. Només has de modificar dues prioritats: la de l'estat inicial i la de cada successor. En totes dues, multiplica la heurística per `pes`.

In [ ]:
# TODO: copia ací a_estrella i canvia-la a a_estrella_ponderada(problema, pes=1).
# Recorda: prioritat = g(n) + pes * h(n).


In [ ]:
# Executa aquesta cel·la després de completar a_estrella_ponderada.
for pes in (0, 1, 2, 5, 100):
    problema = ProblemaRomania(EstatRomania('Arad'), EstatRomania('Bucharest'))
    estat = a_estrella_ponderada(problema, pes)
    print(f'Pes {pes}')
    mostra_ruta(estat)
    print()


# Qüestions

Completa una taula amb la ruta, el cost i el nombre d'estats visitats per a cada pes:

| Pes $w$ | Ruta | Cost | Estats visitats | És òptima? |
| ---: | --- | ---: | ---: | --- |
| 0 | | | | |
| 1 | | | | |
| 2 | | | | |
| 5 | | | | |
| 100 | | | | |

1. Quin valor de $w$ correspon a UCS? I a A* estàndard?
2. A partir de quin pes deixa de trobar el cost òptim 418?
3. Per què augmentar $w$ pot reduir els estats explorats?
4. En quin tipus d'aplicació acceptaries una solució ràpida però no òptima? I en quina no l'acceptaries?